instrumentation/tomcat-7.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/tomcat7/ServerHandlerAdvice.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.javaagent.instrumentation.tomcat7;
+
+import static io.opentelemetry.javaagent.instrumentation.tomcat7.TomcatTracer.tracer;
+
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.Scope;
+import io.opentelemetry.instrumentation.api.servlet.AppServerBridge;
+import io.opentelemetry.javaagent.instrumentation.servlet.v3_0.TagSettingAsyncListener;
+import java.util.concurrent.atomic.AtomicBoolean;
+import net.bytebuddy.asm.Advice;
+import org.apache.coyote.Adapter;
+import org.apache.coyote.Request;
+import org.apache.coyote.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@SuppressWarnings("unused")
+public class ServerHandlerAdvice {
+  private static final Logger log = LoggerFactory.getLogger(ServerHandlerAdvice.class);
+
+  @Advice.OnMethodEnter(suppress = Throwable.class)
+  public static void onEnter(
+      @Advice.This Adapter itsMe,
+      @Advice.Argument(0) Request request,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    Context attachedContext = tracer().getServerContext(request);
+    if (attachedContext != null) {
+      log.debug("Unexpected context found before server handler even started: {}", attachedContext);
+      return;
+    }
+
+    context = tracer().startServerSpan(request, itsMe.getClass().getSimpleName());
+
+    scope = context.makeCurrent();
+  }
+
+  @Advice.OnMethodExit(onThrowable = Throwable.class, suppress = Throwable.class)
+  public static void stopSpan(
+      @Advice.Argument(0) Request request,
+      @Advice.Argument(1) Response response,
+      @Advice.Thrown Throwable throwable,
+      @Advice.Local("otelContext") Context context,
+      @Advice.Local("otelScope") Scope scope) {
+    if (scope == null) {
+      return;
+    }
+    scope.close();
+
+    if (context == null) {
+      return;
+    }
+
+    Throwable throwableToReport =
+        throwable != null ? throwable : AppServerBridge.getUnhandledThrowable(context);
```

Comment:
Why do you need this `AppServerBridge.getUnhandledThrowable`? You can record the exception on SERVER span write when it happened inside servlet instrumentation. You don't need to pass it up.

Reply:
Good point! Looks that this works. Will change.